examples/auction/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+        let (account1, ctx1) = new_account_ctx();
+        verify_bid(account1, &ctx1, amount, &mut bid_map, &mut state, amount);
+
+        // 2nd bid: account1 bids `amount` again
+        // should work even though it's the same amount because account1 simply increases their bid
+        verify_bid(account1, &ctx1, amount, &mut bid_map, &mut state, amount + amount);
+
+        // 3rd bid: second account
+        let (account2, ctx2) = new_account_ctx();
+        verify_bid(account2, &ctx2, winning_amount, &mut bid_map, &mut state, winning_amount);
+
+        // trying to finalize auction that is still active
+        // (specifically, the bid is submitted at the last moment, at the AUCTION_END time)
+        let mut ctx4 = ReceiveContextTest::empty();
+        ctx4.set_metadata_slot_time(Timestamp::from_timestamp_millis(AUCTION_END));
+        let finres: Result<ActionsTree, _> = auction_finalize(&ctx4, &mut state);
+        expect_error(finres, FinalizeError::AuctionStillActive,
+                     "Finalizing auction should fail when it's before auction-end time");
+
+        // finalizing auction
+        let mut ctx5 = new_ctx(owner, owner, AUCTION_END + 1);
+        ctx5.set_self_balance(winning_amount);
+        let finres2: Result<ActionsTree, _> = auction_finalize(&ctx5, &mut state);
+        let actions = finres2.expect("Finalizing auction should work");
+        assert_eq!(actions, ActionsTree::simple_transfer(&owner, winning_amount)
+            .and_then(ActionsTree::simple_transfer(&account1, amount + amount)));
+        assert_eq!(state, State {
+            auction_state: AuctionState::Sold(account2),
+            highest_bid: winning_amount,
+            item: ITEM.as_bytes().to_vec(),
+            expiry: Timestamp::from_timestamp_millis(AUCTION_END),
+            bids: bid_map.clone(), // todo bad
+        });
+
+        // attempting to finalize auction again should fail
+        let finres3: Result<ActionsTree, _> = auction_finalize(&ctx5, &mut state);
+        expect_error(finres3, FinalizeError::AuctionFinalized,
+                     "Finalizing auction a second time should fail");
+
+        // attempting to bid again should fail
+        let res4: Result<ActionsTree, _> = auction_bid(&ctx2, big_amount, &mut state);
+        expect_error(res4, ReceiveError::AuctionFinalized,
+                     "Bidding should fail because the auction is finalized");
+    }
+
+    fn verify_bid(
+        account: AccountAddress,
+        ctx: &ContextTest<ReceiveOnlyDataTest>,
+        amount: Amount,
+        bid_map: &mut BTreeMap<AccountAddress, Amount>,
+        mut state: &mut State,
+        highest_bid: Amount,
+    ) {
+        let res: Result<ActionsTree, _> = auction_bid(ctx, amount, &mut state);
+        res.expect("Bidding should pass");
+        bid_map.insert(account, highest_bid);
+        assert_eq!(*state, dummy_state(highest_bid, bid_map.clone()));
+    }
+
+    #[test]
+    fn test_auction_bid_repeated_bid() {
```

Comment:
It would be useful to add a line or two briefly summarizing the test (the same applies to other tests).